### ISSUE TITLE
Fix nightly build

### ISSUE THREAD

**Issue by ian-r-rose** [cagov/caldata-mdsa-caltrans-pems#323]
The nightly build job [failed](https://california.getdbt.com/deploy/6/projects/10/runs/4165) last night.

**Comment by ian-r-rose** [comment]
@pingpingxiu-DOT-ca-gov I took a look into this, and it seems that something has gone wrong with the detector config and controller config uploads. Yesterday, `raw_prd.db96.detector_config` had  61,869 rows , and today it has 866,342 rows, fourteen times as many.

Sure enough when I take a look at the detector config rows in the table, I see that each detector is repeated fourteen times. This is resulting in duplicate data downstream, and causing the failure in CI here. @pingpingxiu-DOT-ca-gov would you be able to look into what might be going wrong?

**Comment by pingpingxiu-DOT-ca-gov** [comment]
I did some experiments for on-demand crawling, leading to duplicates. I thought downstream already handles duplicates.

If not, I can add dedupe logic on Snowsql part.
Currently, the snow sql logic does not do deduplication.

COPY INTO DB96.DETECTOR_CONFIG FROM (SELECT * FROM @~/detector_config) FILE_FORMAT = (TYPE = PARQUET);

As long as I add de-duplication like below, downstream should not see duplicates.

COPY INTO DB96.DETECTOR_CONFIG FROM (SELECT * FROM @~/detector_config
qualify row_number() over (partition by station_id, TIME_ID order by detector_id) = 1) FILE_FORMAT = (TYPE = PARQUET);

**Comment by ian-r-rose** [comment]
> I did some experiments for on-demand crawling, leading to duplicates. I thought downstream already handles duplicates.

I think that's way over-engineering it for this pipeline. These tables are small, and we were uploading them in a single shot. In fact, I strongly recommend *further decoupling* the config pipelines from the VDS crawler. They are logically separate, and one breaking shouldn't affect the other. These tables aren't doing any specific deduplication right now, and we are trusting the uploads to be correct.

Can you revert the changes so that we are using the old behavior with single-uploads of the table?

In the future, let's create issues to discuss changes like this to identify whether there are any expected changes or breakages. From a development perspective, we shouldn't be making major changes to the production pipelines as an experiment, we should be able to test changes in the development environment without breaking the downstream prod transformations.

**Comment by pingpingxiu-DOT-ca-gov** [comment]
Crawling requests should be allowed to over-send, this is a valid use case
because we can get shorter-than-one-day SLA fresh data, rather than waiting for
a daily schedule, in case we need that freshness.

The over-sent crawling requests would break the downstream, let’s fix that.
We need to apply de-duplication logic to config, just like we did deduplication
for VDS30SEC.

I have not made the change yet so I was going to go over with you tomorrow.

**Comment by ian-r-rose** [comment]
This is not a valid use case for the config tables, where we determined that we would just over-write the table every night. I don't think the config tables should be considered part of the "crawler", they are just tasks that need to be run nightly and overwrite the table.

Everyone's workflow is broken right now due to this change, please revert it and we can discuss more tomorrow.

**Comment by pingpingxiu-DOT-ca-gov** [comment]
@ian-r-rose , this is the data issue. I cannot revert back code to fix that
because it was manually triggering the daily cron job. 

According to what you previously advised, pipeline needs to be *idempotent*. 
The downstream pipeline would break whenever data has duplicates, 
it is not idempotent in my own opinion, that we need to think through. 

@ZhenyuZhu-Caltrans

**Comment by ian-r-rose** [comment]
We previously wrote a sql script that uploaded the full table once a day, and truncated the table beforehand. I do not understand why we cannot do that anymore.

**Comment by pingpingxiu-DOT-ca-gov** [comment]
You're referring to the SNOWSQL query we wrote. 
I'm doing that, everything remain the same. 

It's just that we need to make it duplication-proof.

**Comment by ian-r-rose** [comment]
Previously we had scripts that did the following:

1. `select *` on the source tables, save them to a parquet file
2. Truncate the target table in Snowflake so that we are starting fresh
3. Upload that parquet file to Snowflake.

The tables are small, the parquet files are small. Uploading tons of duplicate rows based on a crawl process over-engineering, and I was trying to avoid an overly complex pipeline for some files that could fit in an email attachment.

You said above that you were doing "some experiments for on-demand crawling". My point here is

1. We should not be doing experiments on the production table without discussion, because it can lead to major breakages. All analytics engineering in the dbt project is currently blocked due to this change.
2. This is a good reason to not have the config table pipeline overly tied to the kafka crawler. The tables are not appropriate for kafka, they are much too small and slowly updating.

**Comment by pingpingxiu-DOT-ca-gov** [comment]
@ian-r-rose , please be assured that making pipeline robust is top priority. 

I just want to make my point clear that if a pipeline cannot handle the out-of-schedule
cronjob execution, that it will break on data duplications resulted from that, we need
to fix it to make it robust. 

@ZhenyuZhu-Caltrans

**Comment by ian-r-rose** [comment]
The previous pipeline should have been robust to being scheduled at any time via cron: the tables are small and slowly changing, and we made the decision to upload them in a single shot every time the script runs. We could run the upload script 100 times a day, and the resulting table should look the same in Snowflake. That's because we took the following approach:

1. `select *` from the source table. There are no duplicates at this stage
2. Save the result as a single parquet file. There are no duplicates at this stage
3. Upload the parquet file to the Snowflake stage.
4. Truncate the target table in Snowflake. It is now empty, and there are no duplicates
5. Copy the parquet file from the stage into the target table. It now has the whole dataset and there are no duplicates.

There is no need to worry about crawlers, or over-sending data, because we are just uploading the whole dataset ever time. Which part of the above is no longer true?
